servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/KeepAliveManager.java

```diff
@@ -0,0 +1,321 @@
+        // write future may not be executed which sets the timer.
+        gracefulCloseState = GRACEFUL_CLOSE_START;
+
+        // The graceful close process is described in [1]. It involves sending 2 GOAWAY frames. The first
+        // GOAWAY has last-stream-id=<maximum stream ID> to indicate no new streams can be created, wait for 2 RTT
+        // time duration for inflight frames to land, and the second GOAWAY includes the maximum known stream ID.
+        // To account for 2 RTTs we can send a PING and when the PING(ACK) comes back we can send the second GOAWAY.
+        // [1] https://tools.ietf.org/html/rfc7540#section-6.8
+        DefaultHttp2GoAwayFrame goAwayFrame = new DefaultHttp2GoAwayFrame(NO_ERROR);
+        goAwayFrame.setExtraStreamIds(Integer.MAX_VALUE);
+        channel.write(goAwayFrame);
+        channel.writeAndFlush(new DefaultHttp2PingFrame(GRACEFUL_CLOSE_PING_CONTENT)).addListener(future -> {
+            // If gracefulCloseState is not GRACEFUL_CLOSE_START that means we have already received the PING(ACK) and
+            // there is no need to apply the timeout.
+            if (future.isSuccess() && gracefulCloseState == GRACEFUL_CLOSE_START) {
+                gracefulCloseState = scheduler.afterMillis(() -> {
+                    // If the PING(ACK) times out we may have under estimated the 2RTT time so we
+                    // optimistically keep the connection open and rely upon higher level timeouts to tear
+                    // down the connection.
+                    LOGGER.debug("channel={} timeout {}ms waiting for PING(ACK) during graceful close.",
+                            channel, pingAckTimeoutMillis);
+                    gracefulCloseWriteSecondGoAway();
+                }, pingAckTimeoutMillis);
+            }
+        });
+    }
+
+    private void gracefulCloseWriteSecondGoAway() {
+        assert channel.eventLoop().inEventLoop();
+
+        if (gracefulCloseState == GRACEFUL_CLOSE_SECOND_GO_AWAY_SENT) {
+            return;
+        }
+
+        gracefulCloseState = GRACEFUL_CLOSE_SECOND_GO_AWAY_SENT;
+
+        channel.writeAndFlush(new DefaultHttp2GoAwayFrame(NO_ERROR)).addListener(future -> {
+            if (activeChildChannels == 0) {
+                close0();
+            }
+        });
+    }
+
+    private void close0() {
+        assert channel.eventLoop().inEventLoop();
+
+        if (gracefulCloseState == CLOSED && keepAliveState == CLOSED) {
+            return;
+        }
+        gracefulCloseState = CLOSED;
+        keepAliveState = CLOSED;
+
+        // The way netty H2 stream state machine works, we may trigger stream closures during writes with flushes
+        // pending behind the writes. In such cases, we may close too early ignoring the writes. Hence we flush before
+        // closure, if there is no write pending then flush is a noop.
+        channel.flush();
+        channel.close();
```

Comment:
should the close be done once all previous written data is really written out on the network ? At the moment we may loose some data if the channel was not writable when flush() is called.

Reply:
(this code is pre-existing, just refactored).

We enter here when `active streams == 0`, there is a fatal error or missing keep-alive. In all these cases we do not expect further graceful closure. 
Practically, this would require we track all writes on the channel from here, ROI of which is arguable here.